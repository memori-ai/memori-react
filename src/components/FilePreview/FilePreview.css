.memori--preview-container {
    min-width: 200px;
    padding: 12px;
    border-radius: 8px;
    margin-bottom: 12px;
    animation: slide-in 0.3s ease;
    background: white;
    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
    transition: all 0.3s ease;
  }

  .memori--preview-container.slide-down {
    animation: slide-down 0.3s ease;
  }

  .memori--absolute-preview {
    position: absolute;
    right: 0;
    bottom: 100%;
    left: 0;
  }

  .memori--message-preview {
    margin: 12px 0;
  }

  .memori--preview-list {
    display: flex;
    flex-direction: column;
    gap: 8px;
  }

  .memori--preview-item {
    position: relative;
    display: flex;
    max-width: fit-content;
    align-items: center;
    padding: 8px;
    border-radius: 8px;
    animation: slide-in 0.3s ease;
    background-color: #f8f9fa;
    cursor: pointer;
    gap: 8px;
    transition: all 0.2s ease;
  }

  .memori--preview-item:hover {
    background-color: #f1f3f5;
    transform: translateX(4px);
  }

  .memori--preview-icon {
    width: 20px;
    height: 20px;
    flex-shrink: 0;
    fill: var(--memori-primary);
  }

  .memori--preview-filename {
    overflow: hidden;
    max-width: 200px;
    color: #495057;
    font-size: 14px;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .memori--modal-title-preview {
    margin-top: 12px;
    margin-bottom: 12px;
    font-size: 20px;
    font-weight: 600;
  }

  .memori--remove-button {
    position: absolute;
    top: -4px;
    right: -4px;
    padding: 4px;
    border: none;
    border-radius: 50%;
    background-color: #e03131;
    color: white;
    cursor: pointer;
    opacity: 0;
    transform: scale(0.75);
    transition: all 0.2s ease;
  }

  .memori--remove-button.visible {
    opacity: 1;
    transform: scale(1);
  }

  .memori--remove-button:hover {
    background-color: #c92a2a;
    transform: scale(1.1);
  }

  .memori--remove-icon {
    width: 12px;
    height: 12px;
  }

  .memori--modal-preview-file {
    width: 100%;
    max-width: 800px;
    height: 100%;
    padding: 12px;
  }

  @keyframes fadeIn {
    from { opacity: 0; }
    to { opacity: 1; }
  }

  @keyframes scaleIn {
    from { 
      opacity: 0;
      transform: scale(0.95);
    }
    to { 
      opacity: 1;
      transform: scale(1);
    }
  }

  @keyframes slide-up {
    from {
      opacity: 0;
      transform: translateY(10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }
  
  @keyframes slide-in {
    from {
      opacity: 0;
      transform: translateY(-10px);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }

  @keyframes slide-down {
    from {
      opacity: 1;
      transform: translateY(0);
    }
  }