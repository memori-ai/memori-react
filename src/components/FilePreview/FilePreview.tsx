import React, { useState } from 'react';
import { PreviewIcon } from '../icons/Preview';
import { DocumentIcon } from '../icons/Document';
import { ImageIcon } from '../icons/Image';
import CloseIcon from '../icons/Close';
import Button from '../ui/Button';
import Modal from '../ui/Modal';
import { stripHTML } from '../../helpers/utils';

type FilePreviewProps = {
  previewFiles: any;
  removeFile: (id: string, mediumID: string | undefined) => void;
  allowRemove?: boolean;
  isMessagePreview?: boolean;
};

const FilePreview = ({
  previewFiles,
  removeFile,
  allowRemove = true,
  isMessagePreview = false,
}: FilePreviewProps) => {
  const [selectedFile, setSelectedFile] = useState<{
    name: string;
    id: string;
    content: string;
    type?: string;
  } | null>(null);
  
  const [hoveredId, setHoveredId] = useState<string | null>(null);

  const getFileType = (filename: string, type?: string) => {
    // If type is explicitly provided, use it first
    if (type === 'image') {
      const extension = filename.split('.').pop()?.toLowerCase();
      switch (extension) {
        case 'jpg':
        case 'jpeg':
          return 'JPEG';
        case 'png':
          return 'PNG';
        default:
          return 'Image';
      }
    }
    
    // Otherwise use extension
    const extension = filename.split('.').pop()?.toLowerCase();
    switch (extension) {
      case 'pdf':
        return 'PDF';
      case 'txt':
        return 'Text';
      case 'json':
        return 'JSON';
      case 'xlsx':
        return 'Excel';
      case 'csv':
        return 'CSV';
      case 'jpg':
      case 'jpeg':
        return 'JPEG';
      case 'png':
        return 'PNG';
      default:
        return 'Document';
    }
  };

  // Detect if the content is an image URL
  const isImageContent = (content: string, type?: string): boolean => {
    if (type === 'image') return true;
    
    // Check if the content has image file extension or is an image URL
    const hasImageExtension = /\.(jpg|jpeg|png|gif|webp|svg)$/i.test(content);
    const isImageUrl = content.startsWith('http') && 
                       (content.includes('/image/') || 
                        content.includes('/img/') || 
                        hasImageExtension);
    
    return isImageUrl || hasImageExtension;
  };

  return (
    <>
      {previewFiles.length > 0 && (
        <div
          className={`memori--preview-container  ${
            isMessagePreview
              ? 'memori--message-preview'
              : 'memori--absolute-preview'
          }`}
        >
          <div className="memori--preview-list">
            {previewFiles.map((file: any) => (
              <div
                key={file.id}
                className={`memori--preview-item ${
                  isImageContent(file.content, file.type) 
                    ? 'memori--preview-item--image' 
                    : 'memori--preview-item--document'
                }`}
                onMouseEnter={() => setHoveredId(file.id)}
                onMouseLeave={() => setHoveredId(null)}
                onClick={() => setSelectedFile(file)}
              >
                {isImageContent(file.content, file.type) ? (
                  <div className="memori--preview-thumbnail">
                    <img src={file.content} alt={file.name} />
                  </div>
                ) : (
                  <DocumentIcon className="memori--preview-icon" />
                )}
                
                <div className="memori--preview-file-info">
                  <span className="memori--preview-filename">{file.name}</span>
                  <span className="memori--preview-filetype">
                    {getFileType(file.name, file.type)}
                  </span>
                </div>
                
                {allowRemove && (
                  <Button
                    shape="rounded"
                    icon={<CloseIcon />}
                    danger
                    className={`memori--remove-button ${
                      hoveredId === file.id ? 'visible' : ''
                    }`}
                    onClick={e => {
                      e.stopPropagation();
                      removeFile(file.id, file?.mediumID);
                    }}
                  />
                )}
              </div>
            ))}
          </div>
        </div>
      )}
      
      <Modal
        width="80%"
        widthMd="80%"
        open={!!selectedFile}
        className="memori--modal-preview-file"
        onClose={() => setSelectedFile(null)}
        closable
        title={selectedFile?.name}
      >
        <div 
          className="memori--preview-content" 
          style={{ 
            maxHeight: '70vh', 
            overflowY: 'auto',
            textAlign: 'center',
            whiteSpace: selectedFile && !isImageContent(selectedFile.content, selectedFile.type) ? 'pre-wrap' : 'normal'
          }}
        >
          {selectedFile && isImageContent(selectedFile.content, selectedFile.type) ? (
            <>
              <img 
                src={selectedFile.content} 
                alt={selectedFile.name} 
                style={{ maxWidth: '100%', maxHeight: '60vh' }} 
            />
            </>
          ) : (
            stripHTML(selectedFile?.content || '')
          )}
        </div>
      </Modal>
    </>
  );
};

export default FilePreview;