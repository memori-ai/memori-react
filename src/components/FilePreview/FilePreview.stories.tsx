import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react';
import I18nWrapper from '../../I18nWrapper';
import FilePreview from './FilePreview';

import './FilePreview.css';

const meta: Meta = {
  title: 'Widget/File Preview',
  component: FilePreview,
  argTypes: {
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
  parameters: {
    controls: { expanded: true },
  },
};


type Props = {
  onUpload: (text: string) => void;
  convertapiToken: string;
};

export default meta;

const Template: Story<Props> = args => {
    const [previewFiles, setPreviewFiles] = useState<{ name: string; id: string; content: string; }[]>([]);

  return (
    <I18nWrapper>
      <div
        style={{
          minHeight: '200px',
          display: 'flex',
          alignItems: 'flex-end',
        }}
      >
        <FilePreview {...args} previewFiles={previewFiles} removeFile={
            (id: string) => {
                setPreviewFiles(previewFiles.filter(file => file.id !== id));
            }
        } />
      </div>
    </I18nWrapper>
  );
};

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default = Template.bind({});
Default.args = {
  onUpload: () => {},
  convertapiToken: '',
};

export const Disabled = Template.bind({});
Disabled.args = {
  onUpload: () => {},
  convertapiToken: '',
  disabled: true,
};
