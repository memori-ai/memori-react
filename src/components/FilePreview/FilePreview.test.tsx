import React from 'react';
import { render } from '@testing-library/react';
import FilePreview from './FilePreview';

it('renders FilePreview unchanged', () => {
  const { container } = render(
    <FilePreview previewFiles={[]} removeFile={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});

it('renders FilePreview with one file', () => {
  const { container } = render(
    <FilePreview previewFiles={[{ name: 'test.pdf', id: '1', content: 'test' }]} removeFile={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});


it('renders FilePreview with two files', () => {
  const { container } = render(
    <FilePreview previewFiles={[{ name: 'test.pdf', id: '1', content: 'test' }, { name: 'test.pdf', id: '2', content: 'test' }]} removeFile={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});

