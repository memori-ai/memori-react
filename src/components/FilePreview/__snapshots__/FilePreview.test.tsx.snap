// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FilePreview unchanged 1`] = `<div />`;

exports[`renders FilePreview with one file 1`] = `
<div>
  <div
    class="memori--preview-container  memori--absolute-preview"
  >
    <div
      class="memori--preview-list"
    >
      <div
        class="memori--preview-item memori--preview-item--document"
      >
        <svg
          class="memori--preview-icon"
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2.5H6C5.60218 2.5 5.22064 2.65804 4.93934 2.93934C4.65804 3.22064 4.5 3.60218 4.5 4V20C4.5 20.3978 4.65804 20.7794 4.93934 21.0607C5.22064 21.342 5.60218 21.5 6 21.5H18C18.3978 21.5 18.7794 21.342 19.0607 21.0607C19.342 20.7794 19.5 20.3978 19.5 20V8L14 2.5Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M14 2.5V8H19.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 13H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 17H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M10 9H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <div
          class="memori--preview-file-info"
        >
          <span
            class="memori--preview-filename"
          >
            test.pdf
          </span>
          <span
            class="memori--preview-filetype"
          >
            PDF
          </span>
        </div>
        <button
          class="memori-button memori-button--rounded memori-button--padded memori-button--icon-only memori-button--danger memori--remove-button "
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                fill="currentColor"
                stroke="none"
                stroke-width="0"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders FilePreview with two files 1`] = `
<div>
  <div
    class="memori--preview-container  memori--absolute-preview"
  >
    <div
      class="memori--preview-list"
    >
      <div
        class="memori--preview-item memori--preview-item--document"
      >
        <svg
          class="memori--preview-icon"
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2.5H6C5.60218 2.5 5.22064 2.65804 4.93934 2.93934C4.65804 3.22064 4.5 3.60218 4.5 4V20C4.5 20.3978 4.65804 20.7794 4.93934 21.0607C5.22064 21.342 5.60218 21.5 6 21.5H18C18.3978 21.5 18.7794 21.342 19.0607 21.0607C19.342 20.7794 19.5 20.3978 19.5 20V8L14 2.5Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M14 2.5V8H19.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 13H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 17H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M10 9H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <div
          class="memori--preview-file-info"
        >
          <span
            class="memori--preview-filename"
          >
            test.pdf
          </span>
          <span
            class="memori--preview-filetype"
          >
            PDF
          </span>
        </div>
        <button
          class="memori-button memori-button--rounded memori-button--padded memori-button--icon-only memori-button--danger memori--remove-button "
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                fill="currentColor"
                stroke="none"
                stroke-width="0"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="memori--preview-item memori--preview-item--document"
      >
        <svg
          class="memori--preview-icon"
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 2.5H6C5.60218 2.5 5.22064 2.65804 4.93934 2.93934C4.65804 3.22064 4.5 3.60218 4.5 4V20C4.5 20.3978 4.65804 20.7794 4.93934 21.0607C5.22064 21.342 5.60218 21.5 6 21.5H18C18.3978 21.5 18.7794 21.342 19.0607 21.0607C19.342 20.7794 19.5 20.3978 19.5 20V8L14 2.5Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M14 2.5V8H19.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 13H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M16 17H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
          <path
            d="M10 9H8"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <div
          class="memori--preview-file-info"
        >
          <span
            class="memori--preview-filename"
          >
            test.pdf
          </span>
          <span
            class="memori--preview-filetype"
          >
            PDF
          </span>
        </div>
        <button
          class="memori-button memori-button--rounded memori-button--padded memori-button--icon-only memori-button--danger memori--remove-button "
        >
          <span
            class="memori-button--icon"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              role="img"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                fill="currentColor"
                stroke="none"
                stroke-width="0"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
